lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -49,6 +49,13 @@ def find(request)
     node = node_from_request(facts, request)
     node.trusted_data = Puppet.lookup(:trusted_information) { Puppet::Context::TrustedInformation.local(node) }.to_h
 
+    if node.environment
+      # TODO Do we want to end up resetting this every time? Do we expect any changes
```

Comment:
Depends on environment caching. This is where the question of whether to reset the text domain should be tied to the environment object.

I would do something like `node.environment.use_domain`, and have that reset_text_domain and load_from_modulepath the first time then use a cached domain for subsequent queries.

Reply:
That sounds like a good solution, I'll look into that.

Reply:
Do you think once that is implemented it will be sufficient for https://tickets.puppetlabs.com/browse/PUP-8013 ?

Reply:
I think so, I'll need to review PUP-8013.

Reply:
The other missing piece is deleting a domain when that environment is destroyed.

Reply:
Is this https://github.com/puppetlabs/puppet/blob/master/lib/puppet/environments.rb#L366-L368 a good place to put that? Or am I misunderstanding how that happens?

Reply:
That's my best guess. Both clear and clear_all.